toolbox/util/github.go

```diff
@@ -0,0 +1,337 @@
+	log.Printf("Gathering all issues from Github for %s/%s. This may take a while...", owner, repo)
+
+	start := time.Now().Round(time.Second)
+
+	lo := &github.ListOptions{
+		Page:    1,
+		PerPage: 100,
+	}
+	ilo := &github.IssueListByRepoOptions{
+		State:       "all",
+		ListOptions: *lo,
+	}
+
+	issues, resp, err := c.Issues.ListByRepo(context.Background(), owner, repo, ilo)
+	if err != nil {
+		return nil, err
+	}
+	RenderProgressBar(ilo.ListOptions.Page, resp.LastPage, time.Now().Round(time.Second).Sub(start).String(), true)
+
+	ilo.ListOptions.Page++
+
+	for ilo.ListOptions.Page <= resp.LastPage {
+		is, _, err := c.Issues.ListByRepo(context.Background(), owner, repo, ilo)
+		if err != nil {
+			// New line following the progress bar
+			fmt.Printf("\n")
+			return nil, err
+		}
+		RenderProgressBar(ilo.ListOptions.Page, resp.LastPage, time.Now().Round(time.Second).Sub(start).String(), false)
+
+		for _, i := range is {
+			issues = append(issues, i)
+		}
+		ilo.ListOptions.Page++
+	}
+	// New line following the progress bar
+	fmt.Printf("\n")
+	log.Printf("All issues fetched.")
+	return issues, nil
+}
+
+// ListAllTags lists all tags for given owner and repo.
+func ListAllTags(c *github.Client, owner, repo string) ([]*github.RepositoryTag, error) {
+	lo := &github.ListOptions{
+		Page:    1,
+		PerPage: 100,
+	}
+
+	tags, resp, err := c.Repositories.ListTags(context.Background(), owner, repo, lo)
+	if err != nil {
+		return nil, err
+	}
+	lo.Page++
+
+	for lo.Page <= resp.LastPage {
+		ta, _, err := c.Repositories.ListTags(context.Background(), owner, repo, lo)
+		if err != nil {
+			return nil, err
+		}
+		for _, t := range ta {
+			tags = append(tags, t)
+		}
+		lo.Page++
+	}
+	return tags, nil
+}
+
+// ListAllCommits lists all commits for given owner, repo, branch and time range.
+func ListAllCommits(c *github.Client, owner, repo, branch string, start, end time.Time) ([]*github.RepositoryCommit, error) {
+	lo := &github.ListOptions{
+		Page:    1,
+		PerPage: 100,
+	}
+
+	clo := &github.CommitsListOptions{
+		SHA:         branch,
+		Since:       start,
+		Until:       end,
+		ListOptions: *lo,
+	}
+
+	commits, resp, err := c.Repositories.ListCommits(context.Background(), owner, repo, clo)
+	if err != nil {
+		return nil, err
+	}
+	clo.ListOptions.Page++
+
+	for clo.ListOptions.Page <= resp.LastPage {
+		co, _, err := c.Repositories.ListCommits(context.Background(), owner, repo, clo)
+		if err != nil {
+			return nil, err
+		}
+		for _, commit := range co {
+			commits = append(commits, commit)
+		}
+		clo.ListOptions.Page++
+	}
+	return commits, nil
+}
+
+// GetCommitDate gets commit time for given tag/commit, provided with repository tags and commits.
+// The function returns ok as false if input tag/commit cannot be found in the repository.
+func GetCommitDate(c *github.Client, owner, repo, tagCommit string, tags []*github.RepositoryTag) (date time.Time, ok bool) {
+	// If input string is a tag, convert it into SHA
```

Comment:
The input `tagCommit` can be either a tag or a SHA. The input `tags` is a list of all the tags in that repo and is used to look up a tag and get the corresponding SHA. This is because github API's git service can only get one commit by SHA (not by tag).

Reply:
I know this is your way, which is the proper way to call this function. But what if people call it with the illegal inputs, say giving you both `tagCommit` and `tags`, or `tags` which are not pointing to the same SHA, you need to tell this issue, log out message or/and throw out the error.
You are not the only one who is going to use this function later on, so it's your responsibility to make sure your function will not misbehavior with some wrong input.

Reply:
You're right. Updated.